
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>Working with the modules on Bianca - UPPMAX for Deep Learning</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.2afb09e1.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:300,300i,400,400i,700,700i%7CJetBrains+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Lato";--md-code-font:"JetBrains Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#working-with-the-modules-on-bianca" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="UPPMAX for Deep Learning" class="md-header__button md-logo" aria-label="UPPMAX for Deep Learning" data-md-component="logo">
      
  <img src="../assets/logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            UPPMAX for Deep Learning
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Working with the modules on Bianca
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../intro/" class="md-tabs__link">
          
  
  
  Course content

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="UPPMAX for Deep Learning" class="md-nav__button md-logo" aria-label="UPPMAX for Deep Learning" data-md-component="logo">
      
  <img src="../assets/logo.svg" alt="logo">

    </a>
    UPPMAX for Deep Learning
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Course content
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Course content
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../hardware_snowy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hardware
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../slurm-intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Slurm
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../isolated_envs/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Isolated Environments
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../jupyter/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Jupyter Notebooks
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../VSCode/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    VS Code
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../DL_exercises/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    DL Exercises
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../profiling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Profling
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#modules" class="md-nav__link">
    <span class="md-ellipsis">
      Modules
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installed-software" class="md-nav__link">
    <span class="md-ellipsis">
      Installed software
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflows" class="md-nav__link">
    <span class="md-ellipsis">
      Workflows
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="working-with-the-modules-on-bianca">Working with the modules on Bianca<a class="headerlink" href="#working-with-the-modules-on-bianca" title="Anchor link to this section for reference">&para;</a></h1>
<div class="admonition info">
<p class="admonition-title">Objectives</p>
<ul>
<li>Using module system commands</li>
<li>Use software in the module system for some typical workflows</li>
</ul>
</div>
<ul>
<li>800+ programs and packages are installed, with usually many versions of each.</li>
<li>To avoid chaos and collisions, they are managed by a <strong>module system</strong> called <a href="https://lmod.readthedocs.io/en/latest/">LMOD</a>.</li>
<li>By default, the only software available on Bianca and other UPPMAX clusters are standard Linux tools.</li>
<li><strong>All other software</strong> is made available by using the module system.</li>
<li>This means users must explicitly request software, together with software versions, that they require.</li>
<li>This turns out to be quite easy to do. It also improves readability and reproducibility.</li>
<li>Nearly all software in the module system is available on all UPPMAX clusters.</li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<ul>
<li>To access bioinformatics tools, load the <strong>bioinfo-tools</strong> module first.</li>
</ul>
</div>
<h3 id="modules">Modules<a class="headerlink" href="#modules" title="Anchor link to this section for reference">&para;</a></h3>
<ul>
<li><a href="https://www.uppmax.uu.se/resources/software/">Software at UPPMAX</a></li>
<li><a href="https://www.uppmax.uu.se/resources/software/module-system/">Module system</a></li>
</ul>
<h4 id="some-commands">Some commands<a class="headerlink" href="#some-commands" title="Anchor link to this section for reference">&para;</a></h4>
<p>The <code>module</code> command is the basic interface to the module system.
The <code>ml</code> shortcut command is also available.</p>
<ul>
<li>list all modules immediately available, or search for a specific available module<ul>
<li><code>module avail</code> or <code>ml av</code></li>
<li><code>module avail *tool*</code> or <code>ml av *tool*</code></li>
</ul>
</li>
</ul>
<p>This command is not so smart, though, especially when searching for a specific tool, or a bioinformatics tool.
It only reports modules that are immediately available.</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>$ ml av R
</code></pre></div>
outputs everything that has an <code>r</code> in the name... not useful.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>$ ml av samtools
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>No module(s) or extension(s) found!
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>Use &quot;module spider&quot; to find all possible modules and extensions.
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>Use &quot;module keyword key1 key2 ...&quot; to search for all possible modules matching any of the &quot;keys&quot;.
</code></pre></div>
<p>It is better to use <code>module spider</code> or <code>ml spider</code>.
If there is an exact match, it reports it first.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>$ ml spider R
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>-------------------------------------------
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>  R:
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>-------------------------------------------
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>     Versions:
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>        R/3.0.2
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>        R/3.2.3
<a id="__codelineno-2-9" name="__codelineno-2-9" href="#__codelineno-2-9"></a>        R/3.3.2
<a id="__codelineno-2-10" name="__codelineno-2-10" href="#__codelineno-2-10"></a>        R/3.4.0
<a id="__codelineno-2-11" name="__codelineno-2-11" href="#__codelineno-2-11"></a>        R/3.4.3
<a id="__codelineno-2-12" name="__codelineno-2-12" href="#__codelineno-2-12"></a>        R/3.5.0
<a id="__codelineno-2-13" name="__codelineno-2-13" href="#__codelineno-2-13"></a>        R/3.5.2
<a id="__codelineno-2-14" name="__codelineno-2-14" href="#__codelineno-2-14"></a>        R/3.6.0
<a id="__codelineno-2-15" name="__codelineno-2-15" href="#__codelineno-2-15"></a>        R/3.6.1
<a id="__codelineno-2-16" name="__codelineno-2-16" href="#__codelineno-2-16"></a>        R/4.0.0
<a id="__codelineno-2-17" name="__codelineno-2-17" href="#__codelineno-2-17"></a>        R/4.0.4
<a id="__codelineno-2-18" name="__codelineno-2-18" href="#__codelineno-2-18"></a>        R/4.1.1
<a id="__codelineno-2-19" name="__codelineno-2-19" href="#__codelineno-2-19"></a>        R/4.2.1
<a id="__codelineno-2-20" name="__codelineno-2-20" href="#__codelineno-2-20"></a>     Other possible modules matches:
<a id="__codelineno-2-21" name="__codelineno-2-21" href="#__codelineno-2-21"></a>        454-dataprocessing  ADMIXTURE  ANTLR  ARCS  ARC_assembler  ARPACK-NG  ART  AdapterRemoval  AlienTrimmer  Amber  AnchorWave  Arlequin  Armadillo  ArrowGrid  Bamsurgeon  BclConverter  BioBakery  BioBakery_data  ...
<a id="__codelineno-2-22" name="__codelineno-2-22" href="#__codelineno-2-22"></a>
<a id="__codelineno-2-23" name="__codelineno-2-23" href="#__codelineno-2-23"></a>-------------------------------------------
<a id="__codelineno-2-24" name="__codelineno-2-24" href="#__codelineno-2-24"></a>  To find other possible module matches execute:
<a id="__codelineno-2-25" name="__codelineno-2-25" href="#__codelineno-2-25"></a>
<a id="__codelineno-2-26" name="__codelineno-2-26" href="#__codelineno-2-26"></a>      $ module -r spider &#39;.*R.*&#39;
<a id="__codelineno-2-27" name="__codelineno-2-27" href="#__codelineno-2-27"></a>
<a id="__codelineno-2-28" name="__codelineno-2-28" href="#__codelineno-2-28"></a>-------------------------------------------
<a id="__codelineno-2-29" name="__codelineno-2-29" href="#__codelineno-2-29"></a>  For detailed information about a specific &quot;R&quot; package (including how to load the modules) use the module&#39;s full name.
<a id="__codelineno-2-30" name="__codelineno-2-30" href="#__codelineno-2-30"></a>  Note that names that have a trailing (E) are extensions provided by other modules.
<a id="__codelineno-2-31" name="__codelineno-2-31" href="#__codelineno-2-31"></a>  For example:
<a id="__codelineno-2-32" name="__codelineno-2-32" href="#__codelineno-2-32"></a>
<a id="__codelineno-2-33" name="__codelineno-2-33" href="#__codelineno-2-33"></a>     $ module spider R/4.2.1
<a id="__codelineno-2-34" name="__codelineno-2-34" href="#__codelineno-2-34"></a>-------------------------------------------
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>$ ml spider samtools
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>-------------------------------------------
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>  samtools:
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>-------------------------------------------
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>     Versions:
<a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a>        samtools/0.1.12-10
<a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a>        samtools/0.1.19
<a id="__codelineno-3-9" name="__codelineno-3-9" href="#__codelineno-3-9"></a>        samtools/1.1
<a id="__codelineno-3-10" name="__codelineno-3-10" href="#__codelineno-3-10"></a>        samtools/1.2
<a id="__codelineno-3-11" name="__codelineno-3-11" href="#__codelineno-3-11"></a>        samtools/1.3
<a id="__codelineno-3-12" name="__codelineno-3-12" href="#__codelineno-3-12"></a>        samtools/1.4
<a id="__codelineno-3-13" name="__codelineno-3-13" href="#__codelineno-3-13"></a>        samtools/1.5_debug
<a id="__codelineno-3-14" name="__codelineno-3-14" href="#__codelineno-3-14"></a>        samtools/1.5
<a id="__codelineno-3-15" name="__codelineno-3-15" href="#__codelineno-3-15"></a>        samtools/1.6
<a id="__codelineno-3-16" name="__codelineno-3-16" href="#__codelineno-3-16"></a>        samtools/1.8
<a id="__codelineno-3-17" name="__codelineno-3-17" href="#__codelineno-3-17"></a>        samtools/1.9
<a id="__codelineno-3-18" name="__codelineno-3-18" href="#__codelineno-3-18"></a>        samtools/1.10
<a id="__codelineno-3-19" name="__codelineno-3-19" href="#__codelineno-3-19"></a>        samtools/1.12
<a id="__codelineno-3-20" name="__codelineno-3-20" href="#__codelineno-3-20"></a>        samtools/1.14
<a id="__codelineno-3-21" name="__codelineno-3-21" href="#__codelineno-3-21"></a>        samtools/1.16
<a id="__codelineno-3-22" name="__codelineno-3-22" href="#__codelineno-3-22"></a>        samtools/1.17
<a id="__codelineno-3-23" name="__codelineno-3-23" href="#__codelineno-3-23"></a>     Other possible modules matches:
<a id="__codelineno-3-24" name="__codelineno-3-24" href="#__codelineno-3-24"></a>        SAMtools
<a id="__codelineno-3-25" name="__codelineno-3-25" href="#__codelineno-3-25"></a>
<a id="__codelineno-3-26" name="__codelineno-3-26" href="#__codelineno-3-26"></a>-------------------------------------------
<a id="__codelineno-3-27" name="__codelineno-3-27" href="#__codelineno-3-27"></a>  To find other possible module matches execute:
<a id="__codelineno-3-28" name="__codelineno-3-28" href="#__codelineno-3-28"></a>
<a id="__codelineno-3-29" name="__codelineno-3-29" href="#__codelineno-3-29"></a>      $ module -r spider &#39;.*samtools.*&#39;
<a id="__codelineno-3-30" name="__codelineno-3-30" href="#__codelineno-3-30"></a>
<a id="__codelineno-3-31" name="__codelineno-3-31" href="#__codelineno-3-31"></a>-------------------------------------------
<a id="__codelineno-3-32" name="__codelineno-3-32" href="#__codelineno-3-32"></a>  For detailed information about a specific &quot;samtools&quot; package (including how to load the modules) use the module&#39;s full name.
<a id="__codelineno-3-33" name="__codelineno-3-33" href="#__codelineno-3-33"></a>  Note that names that have a trailing (E) are extensions provided by other modules.
<a id="__codelineno-3-34" name="__codelineno-3-34" href="#__codelineno-3-34"></a>  For example:
<a id="__codelineno-3-35" name="__codelineno-3-35" href="#__codelineno-3-35"></a>
<a id="__codelineno-3-36" name="__codelineno-3-36" href="#__codelineno-3-36"></a>     $ module spider samtools/1.17
<a id="__codelineno-3-37" name="__codelineno-3-37" href="#__codelineno-3-37"></a>-------------------------------------------
</code></pre></div>
<p>The final bit of output tells us more about a specific module version, including the special step required to access all bioinformatics modules.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>$ ml spider samtools/1.17
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>-------------------------------------------
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>  samtools: samtools/1.17
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a>-------------------------------------------
<a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a>
<a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a>    You will need to load all module(s) on any one of the lines below before the &quot;samtools/1.17&quot; module is available to load.
<a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a>
<a id="__codelineno-4-9" name="__codelineno-4-9" href="#__codelineno-4-9"></a>      bioinfo-tools
<a id="__codelineno-4-10" name="__codelineno-4-10" href="#__codelineno-4-10"></a>
<a id="__codelineno-4-11" name="__codelineno-4-11" href="#__codelineno-4-11"></a>    Help:
<a id="__codelineno-4-12" name="__codelineno-4-12" href="#__codelineno-4-12"></a>        samtools - use samtools 1.17
<a id="__codelineno-4-13" name="__codelineno-4-13" href="#__codelineno-4-13"></a>
<a id="__codelineno-4-14" name="__codelineno-4-14" href="#__codelineno-4-14"></a>        Version 1.17
</code></pre></div>
<p>This reminds us that we need to load the <code>bioinfo-tools</code> module to be able to load <code>samtools/1.17</code>.
Again, this is required (just once) before loading bioinformatics software.</p>
<ul>
<li>Load a module <ul>
<li><code>module load &lt;module name&gt;</code> or <code>ml &lt;module name&gt;</code></li>
</ul>
</li>
</ul>
<p>When loading a module, there is a "default" module available, which is almost always the latest version.
However, we rarely want to rely on that.
For reproducibility, we want to load specific version of our bioinformatics tools.
To load the <code>samtools/1.17</code> module, which is a bioinformatics module.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>$ ml bioinfo-tools
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>$ ml samtools/1.17
</code></pre></div>
<ul>
<li>List the loaded modules<ul>
<li><code>module list</code> or simply <code>ml</code></li>
</ul>
</li>
</ul>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>$ ml
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>Currently Loaded Modules:
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>  1) uppmax   2) bioinfo-tools   3) samtools/1.17
</code></pre></div>
<p>To load <code>GATK/4.3.0.0</code> now, <code>bioinfo-tools</code> is not required because it is already loaded.
Loading this module also shows that sometimes, loading a module results in a message that is helpful for using the module at UPPMAX.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>$ ml GATK/4.3.0.0
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>Note that all versions of GATK starting with 4.0.8.0 use a different wrapper
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a>script (gatk) than previous versions of GATK.  You might need to update your
<a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a>jobs accordingly.
<a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a>
<a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a>The complete GATK resource bundle is in /sw/data/GATK
<a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a>
<a id="__codelineno-7-8" name="__codelineno-7-8" href="#__codelineno-7-8"></a>See &#39;module help GATK/4.3.0.0&#39; for information on activating the GATK conda
<a id="__codelineno-7-9" name="__codelineno-7-9" href="#__codelineno-7-9"></a>environment for using DetermineGermlineContigPloidy and similar other tools.
</code></pre></div>
<p>This message references the command <code>module help GATK/4.3.0.0</code> for additional help with this module.
All modules have at least a brief help message. Some (such as GATK/4.3.0.0) have more extensive help that guides users using features of the modules at UPPMAX.
It is <em>not</em> general help for using the tool itself.</p>
<ul>
<li>Display a brief module-specific help.<ul>
<li><code>module help &lt;module name&gt;</code> or <code>ml help &lt;module name&gt;</code> </li>
</ul>
</li>
</ul>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>$ ml help GATK/4.3.0.0
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>-------------- Module Specific Help for &quot;GATK/4.3.0.0&quot; ---------------
<a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a>GATK - use GATK 4.3.0.0
<a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a>Version 4.3.0.0
<a id="__codelineno-8-6" name="__codelineno-8-6" href="#__codelineno-8-6"></a>
<a id="__codelineno-8-7" name="__codelineno-8-7" href="#__codelineno-8-7"></a>**GATK 4.3.0.0**
<a id="__codelineno-8-8" name="__codelineno-8-8" href="#__codelineno-8-8"></a>
<a id="__codelineno-8-9" name="__codelineno-8-9" href="#__codelineno-8-9"></a>Usage:
<a id="__codelineno-8-10" name="__codelineno-8-10" href="#__codelineno-8-10"></a>
<a id="__codelineno-8-11" name="__codelineno-8-11" href="#__codelineno-8-11"></a>    gatk --help     for general options, including how to pass java options
<a id="__codelineno-8-12" name="__codelineno-8-12" href="#__codelineno-8-12"></a>
<a id="__codelineno-8-13" name="__codelineno-8-13" href="#__codelineno-8-13"></a>    gatk --list     to list available tools
<a id="__codelineno-8-14" name="__codelineno-8-14" href="#__codelineno-8-14"></a>
<a id="__codelineno-8-15" name="__codelineno-8-15" href="#__codelineno-8-15"></a>    gatk ToolName -OPTION1 value1 -OPTION2 value2 ...
<a id="__codelineno-8-16" name="__codelineno-8-16" href="#__codelineno-8-16"></a>                  to run a specific tool, e.g., HaplotypeCaller, GenotypeGVCFs, ...
<a id="__codelineno-8-17" name="__codelineno-8-17" href="#__codelineno-8-17"></a>
<a id="__codelineno-8-18" name="__codelineno-8-18" href="#__codelineno-8-18"></a>For more help getting started, see
<a id="__codelineno-8-19" name="__codelineno-8-19" href="#__codelineno-8-19"></a>
<a id="__codelineno-8-20" name="__codelineno-8-20" href="#__codelineno-8-20"></a>    https://software.broadinstitute.org/gatk/documentation/article.php?id=9881
<a id="__codelineno-8-21" name="__codelineno-8-21" href="#__codelineno-8-21"></a>
<a id="__codelineno-8-22" name="__codelineno-8-22" href="#__codelineno-8-22"></a>...
</code></pre></div>
<p>When we list the modules loaded with <code>ml</code>, we see that <code>GATK/4.3.0.0</code> is now loaded, as is its prerequisite module <code>java/sun_jdk1.8.0_151</code>.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>$ ml
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a>Currently Loaded Modules:
<a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a>  1) uppmax   2) bioinfo-tools   3) samtools/1.17   4) java/sun_jdk1.8.0_151   5) GATK/4.3.0.0
</code></pre></div>
<p>Modules can also be unloaded, which also unloads their prerequisites.</p>
<ul>
<li>Unload a module <ul>
<li><code>module unload &lt;module name&gt;</code> or <code>ml -&lt;module name&gt;</code></li>
</ul>
</li>
</ul>
<h3 id="installed-software">Installed software<a class="headerlink" href="#installed-software" title="Anchor link to this section for reference">&para;</a></h3>
<p>You can find almost all installed software at:</p>
<ul>
<li><a href="https://www.uppmax.uu.se/resources/software/installed-software/">https://www.uppmax.uu.se/resources/software/installed-software/</a></li>
</ul>
<h4 id="installed-databases">Installed databases<a class="headerlink" href="#installed-databases" title="Anchor link to this section for reference">&para;</a></h4>
<p>You can find descriptions of almost all installed databases at:</p>
<ul>
<li><a href="https://www.uppmax.uu.se/resources/databases/">Installed databases at UPPMAX</a></li>
</ul>
<h3 id="workflows">Workflows<a class="headerlink" href="#workflows" title="Anchor link to this section for reference">&para;</a></h3>
<details class="question" open="open">
<summary>Hands on: Processing a BAM file to a VCF using GATK, and annotating the variants with snpEff</summary>
<p>This workflow uses a pre-made BAM file that contains a subset of reads from a sample from European Nucleotide Archive project <a href="https://www.ebi.ac.uk/ena/browser/view/PRJEB6463">PRJEB6463</a> aligned to human genome build hg38. These reads are from the region <code>chr1:100300000-100800000</code>.</p>
<ol>
<li>
<p>Copy example BAM file to your working directory.
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>$ cp -a /proj/sens2023531/workshop/data/ERR1252289.subset.bam .
</code></pre></div></p>
</li>
<li>
<p>Take a quick look at the BAM file. First see if <code>samtools</code> is available.
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>$ which samtools
</code></pre></div></p>
</li>
<li>
<p>If <code>samtools</code> is not found, load <code>bioinfo-tools</code> then <code>samtools/1.17</code>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>$ ml bioinfo-tools samtools/1.17
</code></pre></div></p>
</li>
<li>
<p>Now create an index for the BAM file, and examine the first 10 reads aligned within the BAM file.
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>$ samtools index ERR1252289.subset.bam
<a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a>$ samtools view ERR1252289.subset.bam | head
</code></pre></div></p>
</li>
<li>
<p>Looks good. Now load the <code>GATK/4.3.0.0</code> module.
<div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>$ module load GATK/4.3.0.0
</code></pre></div></p>
</li>
<li>
<p>Make symbolic links to hg38 genome resources already available on UPPMAX. This provides local symbolic links for the hg38 resources <code>genome.fa</code>, <code>genome.fa.fai</code> and <code>genome.dict</code>.
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>$ ln -s /sw/data/iGenomes/Homo_sapiens/UCSC/hg38/Sequence/WholeGenomeFasta/genome.* .
</code></pre></div></p>
</li>
<li>
<p>Create a VCF containing inferred variants. Speed it up by confining the analysis to this region of chr1.
<div class="highlight"><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>$ gatk HaplotypeCaller --reference genome.fa --input ERR1252289.subset.bam --intervals chr1:100300000-100800000 --output ERR1252289.subset.vcf
</code></pre></div>
This produces as its output the files <code>ERR1252289.subset.vcf</code> and <code>ERR1252289.subset.vcf.idx</code>.</p>
</li>
<li>
<p>Now use <code>snpEff/5.1</code> to annotate the variants. Loading <code>snpEff/5.1</code> results in a change of java prerequisite. Also take a quick look at the help for the module for help with running this tool at UPPMAX.
<div class="highlight"><pre><span></span><code><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a>$ ml snpEff/5.1
<a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a>
<a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a>The following have been reloaded with a version change:
<a id="__codelineno-17-4" name="__codelineno-17-4" href="#__codelineno-17-4"></a>  1) java/sun_jdk1.8.0_151 =&gt; java/OpenJDK_12+32
<a id="__codelineno-17-5" name="__codelineno-17-5" href="#__codelineno-17-5"></a>
<a id="__codelineno-17-6" name="__codelineno-17-6" href="#__codelineno-17-6"></a>$ ml help snpEff/5.1
<a id="__codelineno-17-7" name="__codelineno-17-7" href="#__codelineno-17-7"></a>
<a id="__codelineno-17-8" name="__codelineno-17-8" href="#__codelineno-17-8"></a>------------------- Module Specific Help for &quot;snpEff/5.1&quot; --------------------
<a id="__codelineno-17-9" name="__codelineno-17-9" href="#__codelineno-17-9"></a>    snpEff - use snpEff 5.1
<a id="__codelineno-17-10" name="__codelineno-17-10" href="#__codelineno-17-10"></a>    Version 5.1
<a id="__codelineno-17-11" name="__codelineno-17-11" href="#__codelineno-17-11"></a>
<a id="__codelineno-17-12" name="__codelineno-17-12" href="#__codelineno-17-12"></a>    Usage: java -jar $SNPEFF_ROOT/snpEff.jar ...
<a id="__codelineno-17-13" name="__codelineno-17-13" href="#__codelineno-17-13"></a>
<a id="__codelineno-17-14" name="__codelineno-17-14" href="#__codelineno-17-14"></a>    Usage: java -jar $SNPEFF_ROOT/SnpSift.jar ...
<a id="__codelineno-17-15" name="__codelineno-17-15" href="#__codelineno-17-15"></a>    along with the desired command and possible java options for memory, etc
<a id="__codelineno-17-16" name="__codelineno-17-16" href="#__codelineno-17-16"></a>
<a id="__codelineno-17-17" name="__codelineno-17-17" href="#__codelineno-17-17"></a>    Note that databases must be added by an admin -- request via support@uppmax.uu.se
<a id="__codelineno-17-18" name="__codelineno-17-18" href="#__codelineno-17-18"></a>    See http://snpeff.sourceforge.net/protocol.html for general help
<a id="__codelineno-17-19" name="__codelineno-17-19" href="#__codelineno-17-19"></a>
<a id="__codelineno-17-20" name="__codelineno-17-20" href="#__codelineno-17-20"></a>Every database that is provided by snpEff/5.1 as of this installation is installed.  This complete list
<a id="__codelineno-17-21" name="__codelineno-17-21" href="#__codelineno-17-21"></a>can be generated with
<a id="__codelineno-17-22" name="__codelineno-17-22" href="#__codelineno-17-22"></a>
<a id="__codelineno-17-23" name="__codelineno-17-23" href="#__codelineno-17-23"></a>    java -jar $SNPEFF_ROOT/snpEff.jar databases
<a id="__codelineno-17-24" name="__codelineno-17-24" href="#__codelineno-17-24"></a>
<a id="__codelineno-17-25" name="__codelineno-17-25" href="#__codelineno-17-25"></a>Three additional databases have been installed.
<a id="__codelineno-17-26" name="__codelineno-17-26" href="#__codelineno-17-26"></a>
<a id="__codelineno-17-27" name="__codelineno-17-27" href="#__codelineno-17-27"></a>    Database name                  Description                                      Notes
<a id="__codelineno-17-28" name="__codelineno-17-28" href="#__codelineno-17-28"></a>    -------------                  -----------                                      -----
<a id="__codelineno-17-29" name="__codelineno-17-29" href="#__codelineno-17-29"></a>    c_elegans.PRJNA13758.WS283     Caenorhabditis elegans genome version WS283      MtDNA uses Invertebrate_Mitochondrial codon table
<a id="__codelineno-17-30" name="__codelineno-17-30" href="#__codelineno-17-30"></a>    canFam4.0                      Canis familiaris genome version 4.0
<a id="__codelineno-17-31" name="__codelineno-17-31" href="#__codelineno-17-31"></a>    fAlb15.e73                     Ficedula albicollis ENSEMBLE 73 release
<a id="__codelineno-17-32" name="__codelineno-17-32" href="#__codelineno-17-32"></a>
<a id="__codelineno-17-33" name="__codelineno-17-33" href="#__codelineno-17-33"></a>The complete list of locally installed databases is available at $SNPEFF_ROOT/data/databases_list.installed
<a id="__codelineno-17-34" name="__codelineno-17-34" href="#__codelineno-17-34"></a>
<a id="__codelineno-17-35" name="__codelineno-17-35" href="#__codelineno-17-35"></a>To add your own snpEff database, see the guide at http://pcingola.github.io/SnpEff/se_buildingdb/#option-1-building-a-database-from-gtf-files
</code></pre></div></p>
</li>
<li>
<p>Annotate the variants.
<div class="highlight"><pre><span></span><code><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a>$ java -jar $SNPEFF_ROOT/snpEff.jar eff hg38 ERR1252289.subset.vcf &gt; ERR1252289.subset.snpEff.vcf
</code></pre></div></p>
</li>
<li>
<p>Take a quick look!
<div class="highlight"><pre><span></span><code><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a>$ less ERR1252289.subset.snpEff.vcf
</code></pre></div></p>
</li>
<li>
<p>Compress the annotated VCF and index it, using <code>bgzip</code> and <code>tabix</code> provided by the <code>samtools/1.17</code> module, already loaded.
<div class="highlight"><pre><span></span><code><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>$ bgzip ERR1252289.subset.snpEff.vcf
<a id="__codelineno-20-2" name="__codelineno-20-2" href="#__codelineno-20-2"></a>$ tabix -p vcf ERR1252289.subset.snpEff.vcf.gz
</code></pre></div></p>
</li>
</ol>
</details>
<details class="question" open="open">
<summary>Hands on: Running R within RStudio, use ggplot2 from R_packages/4.1.1</summary>
<ol>
<li>
<p>Load the <code>R_packages/4.1.1</code> module and the latest <code>RStudio</code> module, and start RStudio with <code>rstudio &amp;</code>.
<img alt="ThinLinc load R_packages RStudio" src="../img/modules-1-ml-rstudio.png" /></p>
</li>
<li>
<p>Load the <code>ggplot2</code> R library, provided by <code>R_packages/4.1.1</code>, and produce an example plot.
<img alt="ThinLinc ggplot2" src="../img/modules-2-ggplot2.png" /></p>
</li>
<li>
<p>Save the plot using <code>ggsave</code>.
<img alt="ThinLinc ggsave" src="../img/modules-3-ggsave.png" /></p>
</li>
</ol>
</details>
<details class="question" open="open">
<summary>Hands on: Loading the conda/latest module</summary>
<ol>
<li>Load the <code>conda/latest</code> module.
<div class="highlight"><pre><span></span><code><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>$ ml conda/latest
<a id="__codelineno-21-2" name="__codelineno-21-2" href="#__codelineno-21-2"></a>The variable CONDA_ENVS_PATH contains the location of your environments. Set it to your project&#39;s environments folder if you have one.
<a id="__codelineno-21-3" name="__codelineno-21-3" href="#__codelineno-21-3"></a>Otherwise, the default is ~/.conda/envs. Remember to export the variable with export CONDA_ENVS_PATH=/proj/...
<a id="__codelineno-21-4" name="__codelineno-21-4" href="#__codelineno-21-4"></a>
<a id="__codelineno-21-5" name="__codelineno-21-5" href="#__codelineno-21-5"></a>You may run &quot;source conda_init.sh&quot; to initialise your shell to be able
<a id="__codelineno-21-6" name="__codelineno-21-6" href="#__codelineno-21-6"></a>to run &quot;conda activate&quot; and &quot;conda deactivate&quot; etc.
<a id="__codelineno-21-7" name="__codelineno-21-7" href="#__codelineno-21-7"></a>Just remember that this command adds stuff to your shell outside the scope of the module system.
<a id="__codelineno-21-8" name="__codelineno-21-8" href="#__codelineno-21-8"></a>
<a id="__codelineno-21-9" name="__codelineno-21-9" href="#__codelineno-21-9"></a>REMEMBER TO USE &#39;conda clean -a&#39; once in a while
</code></pre></div></li>
</ol>
<p>We want to set the <code>CONDA_ENVS_PATH</code> variable to a directory within our project, rather than use the default which is our home directory.
If you do not set this variable, your home directory will easily exceed its quotas when creating even a single conda environment.
This will be covered in more detail in the afternoon.</p>
</details>
<div class="admonition abstract">
<p class="admonition-title">keypoints</p>
<ul>
<li>Use the module system to use centrally installed software that is available on all nodes. </li>
<li>Include versions when loading modules, for reproducibility!</li>
<li>Your own installed software, scripts, python packages etc. are available from their paths.</li>
</ul>
</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.path", "navigation.footer", "search.suggest", "content.code.annotate", "content.tabs.link"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>